docs(dt): prepare for R24-03 (#654)
eclipse-tractusx/eclipse-tractusx.github.io: docs-kits/kits/Digital Twin Kit/Software Development View/page_interaction-patterns.md

---
id: dt-kit-interaction-patterns
title: Interaction Patterns
description: 'Digital Twin KIT - Interaction Patterns'
sidebar_position: 2
---

## Interaction Patterns for Distributed Digital Twins

The Catena-X standard CX-0002 (Digital Twins in Catena-X) defines a subset of the

### 1. Fetching a supplier's Twin

| Scenario                                                                    | Participants                                           | Assumptions                                                                 | Links to Use-Cases                                                                               |
|-----------------------------------------------------------------------------|--------------------------------------------------------|-----------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------|

> | The supplier has data required by the OEM. For instance a Bill of Material. | - Data Provider (Supplier) <br/> - Data Consumer (OEM) | 1. Consumer knows Supplier's BPN <br/> 2. Consumer knows an id of the asset | - Industry Core (coming soon) <br/> - [Product Carbon Footprint Kit](../../PCF%20Exchange%20Kit) |

Reviewer: The BOM is not a good example. The OEM has a BOM and this is how the OEM knows whom to ask for data. Example data would be PCF.

> The most common integration pattern with Digital Twins aims to offer asset-related data in expectable ways along the

Reviewer: remove "in expactable ways" - what would be an unexpactalbe way to get data?

> supply chain. This scenario assumes that the supplier offers data about an asset that was sold to an OEM further up the

Reviewer: remove "was sold", substitute with "requested by an OEM (and with legal negotiations being performed beforehand) or similar

chain. Examples for this may include Bills of Material, carbon footprints, technical specifications or regulatory 
documents. An OEM - or really any interested and privileged third party - can assume that a part's supplier offers a Digital
Twin if the data-exchange is based on the Digital Twin Kit (and by proxy CX-0002). The data provider deploys at least:
- Digital Twin Registry (DTR)
  - registered in a DSP-conformant connector offering the DTR to the Dataspace.

>   - registers Digital Twins (ShellDescriptors in AAS vocabulary) with specificAssetIds including at least one that's known

Reviewer: AssetAdministrationShellDescriptors if you want to be precise. There is not "ShellDescriptor".

Reviewer: with asset IDs (global asset ID and specific asset IDs)

  at supplier and customer.
- Submodels
    - registered in a DSP-conformant connector to the Dataspace.

>     - registered to the correct ShellDescriptor in the DTR.

Reviewer: see above: ShellDesriptor is not an official abbreviation.

>     - Serving arbitrary but well-specified json from the `/$value` endpoint

Reviewer: arbitrary json? what does this mean?

> The Data Consumer's only responsibility is to deploy a connector and fetch the data from the right offers.

Reviewer: assuming again, the data consumer has the contracts signed and has the credentials needed

A Bill of Material ([like this one](https://github.com/eclipse-tractusx/sldt-semantic-models/blob/main/io.catenax.single_level_bom_as_built/2.0.0/gen/SingleLevelBomAsBuilt.json))
for instance returns a list of `childItems` that hold for each subcomponent the information that's assumed to be present
for the start of this very scenario (see Assumptions 1, 2). Thus, BoM and the DT deployment in combination 

> enable recursive browsing of the parts-tree, collecting data along the way.

Reviewer: This is true but not the recommended way. The strategy of Catena-X is "One level up, one level down". At least it is not always the same data consumer along the value chain.

### 2. Adding a new Twin for a given Asset

| Scenario                                                                                                      | Participants                                                   | Assumptions                             | Links to Use-Cases    |
|---------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------|-----------------------------------------|-----------------------|
| A party (who is not the supplier of an part) publishes data about a previously uncovered aspect of the asset. | - Data Provider (Supplier) <br/> - Data Provider (Third Party) | 1. Third Party knows an id of the asset | - PURIS (coming soon) |

Let's suppose, a car has reached it end-of-life. The object is sold to a dismantler who disassembles the vehicle and wants

> to publish that exact information. As the Digital Twin is owned by the OEM of the vehicle, it will not be accessible to

Reviewer: The owner decides who is allowed to register new submodels to a digital twin. Typically it is the owner him or herself who is registering submodels only. Additionally, there might be legal or architectural reasons to copy/past submodels from another party. This is why it is possible that one and the same asset (identified via its global asset ID) may have more than one digital twin.

everyone in the network. Thus, the dismantler must publish the data themselves, by deploying:

- Digital Twin Registry (DTR)
    - registered in a DSP-conformant connector offering the DTR to the Dataspace.

>     - registers Digital Twins (ShellDescriptors in AAS vocabulary) with specificAssetIds including at least one that's

Reviewer: see above, ShellDescriptor unknown abbr.

  equivalent to one attached to the OEM's twin.
- Submodel (like [_Certificate of Destruction_](https://github.com/eclipse-tractusx/sldt-semantic-models/blob/4889d8482fc6b233eb0f56f3ac94c5ea4004bc05/io.catenax.certificate_of_destruction/1.0.1/gen/CertificateOfDestruction.json))
    - registered in a DSP-conformant connector to the Dataspace.
    - registered to the correct ShellDescriptor in the DTR.

>     - Serving arbitrary but well-specified json via `GET` requests from the `/$value` endpoint.

Reviewer: see above, since copy/paste

#### 2.1. Discovery by Convention

In certain use-cases, participants can agree on conventions where a specific Submodel may be expected.

> This information can only be provided by the buyer of a certain part who creates a new Twin for the part with identical

Reviewer: not clear. the only asset ID that needs to be identical is the global asset ID.

`assetIds`. Such conventions about how to locate certain types of Submodels should be explicitly mentioned in the 

> Standards and Kits for a use-case as the default hypothesis is always that a [Supplier holds relevant data](#1-fetching-a-suppliers-twin).

Author: added

#### 2.2. Discovery by Registration

> Additionally, the dismantler must signal in a [BPN-Discovery-Service](../page_software-operation-view.md) that he indeed has data on said vehicle.

Reviewer: he or she

Reviewer: Via the BPN Discovery Service it is not possible to decide whether the dismantler has information on a specific vehicle. It is only possible to find the BPN of the dismantler so that in a subsequent steps the data consumer can look for twins in the Digital Twin Registry of the Dismantler for this specific vehicle.

Author: Right, but the dismantler would not register with a certain assetId if they didn't have data on said asset. For all cars the dismantler does not have data on, they don't register at the BPN Discovery Service.

may define conventions like "Data Consumers can assume there to be separate twins for the same asset with the supplier
AND the customer." This however must be explicitly stated to find adoption in Consumers' applications.

### 3. Updating an existing Submodel

| Scenario                                                                                           | Participants                                                   | Assumptions                             | Links to Use-Cases      |
|----------------------------------------------------------------------------------------------------|----------------------------------------------------------------|-----------------------------------------|-------------------------|
| A party (who is not the supplier of an part) updates data already covered by an existing Submodel. | - Data Provider (Supplier) <br/> - Data Provider (Third Party) | 1. Third Party knows an id of the asset | - DCM-AAS (coming soon) |

There may be a scenario where a Supplier deploys a Digital Twin that holds data that will have to be updated during the 

> lifecycle. A potential use-case would be the update of a particular vehicle's BoM - for instance if the engine is replaced.

Reviewer: the BOM as maintained

Author: added

#### 3.2. Access Control in the backend
Depending on the backend's capabilities to fine-tune access control, it may be more beneficial to let the backend (like
a Submodel Repository) decide if data will certain API-operations will be allowed. In this example, it would be possible
to let a user update certain properties in a Submodel but only read others. This differentiation can't be made by the 
connector as a connector will usually be oblivious to the data model and API structure of backend systems like a DTR or
Submodel Repository.

### 4. Updating an existing Twin

| Scenario                                                                  | Participants                                                   | Assumptions                             | Links to Use-Cases |
|---------------------------------------------------------------------------|----------------------------------------------------------------|-----------------------------------------|--------------------|
| A new customer wants their customerPartId as specificAssetId on the twin. | - Data Provider (Supplier) <br/> - Data Provider (Third Party) | 1. Third Party knows an id of the asset | None yet           |

> An update to a twin is any change to the ShellDescriptor via the APIs of the AssetAdministrationShellServiceSpecification

Reviewer: This use case is about changing the submodel that is registered in the Registry for this twin. So the registry information about this twin is not changed at all. Only if there is version information of the submodel contained in the registry (but this is optional)

Author: it's about arbitrary changes to a shell-descriptor. Adding submodels when onboarding use-cases, changing submodel-interfaces in migration scenario...

> or DiscoveryServiceSpecification of AAS Pt 2. While the relevant write-APIs are not mandatory in Catena-X, Data Providers

Reviewer: Pt is not a valid abbr. for part

---
id: dt-kit-interaction-patterns
title: Interaction Patterns
description: 'Digital Twin KIT - Interaction Patterns'
sidebar_position: 2
---

## Interaction Patterns for Distributed Digital Twins

The Catena-X standard CX-0002 (Digital Twins in Catena-X) defines a subset of the AAS-standard that is relevant for 
Catena-X. However, it is concerned only with the network-facing APIs which in some cases may have to be augmented to 

> cover use-case requirements.

Author: elaborated.

While the assumptions are explicitly listed, all patterns assume that both partners involved are properly onboarded to
the Catena-X dataspace and thus have their connectors registered with the EDC Discovery Service like so:

> ```mermaid

Reviewer: why are we explaining this diagram? Nothing specific to Twins in there... ==> remove

Author: It's not specific to twins but relevant. That's why the API is part of this kit.

|-----------------------------------------------------------------------------|--------------------------------------------------------|-----------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------|
| The supplier has data required by the OEM. For instance a Product Passport. | - Data Provider (Supplier) <br/> - Data Consumer (OEM) | 1. Consumer knows Supplier's BPN <br/> 2. Consumer knows an id of the asset | - Industry Core (coming soon) <br/> - [Product Carbon Footprint Kit](../../PCF%20Exchange%20Kit) |

The most common integration pattern with Digital Twins aims to offer asset-related data in expectable ways along the
supply chain. It defines conventions that allow locating data in the network. This most basic scenario assumes that the 
supplier offers data about an asset that was sold to an OEM further up the chain. Supplier and OEM have agreed on 

> contractual terms on which to share data about said asset. Possible examples may include Bills of Material, carbon 

Reviewer: ```suggestion
contractual terms on which data to share about said asset. Possible examples may include Bill of Material, carbon 
```

Author: This suggestion changes the meaning of the sentence. It's about the contractual terms, not the data.

footprints, technical specifications or regulatory documents. An OEM - or really any interested and privileged third 
party - can assume that a part's supplier offers a Digital Twin if the data-exchange is based on the Digital Twin Kit 
(and by proxy CX-0002). The data provider deploys at least:
- Digital Twin Registry (DTR)

>   - registered in a DSP-conformant connector offering the DTR to the Dataspace.

Reviewer: abbreviation DSP not yet introduced. What is it?

>   - registers Digital Twins (AssetAdministrationAAS-descriptors - or AAS descriptor - in AAS vocabulary) with assetIds 

Reviewer: ```suggestion
  - registers Digital Twins (AssetAdministrationShell-descriptors - or AAS descriptor - in AAS vocabulary) with assetIds 
```

>   - registers Digital Twins (AssetAdministrationAAS-descriptors - or AAS descriptor - in AAS vocabulary) with assetIds 

Reviewer: ```suggestion
  - registers Digital Twins (AssetAdministrationAAS-descriptors - or AAS descriptor - in AAS vocabulary) with assetI IDs
```
```suggestion
  - registers Digital Twins (AssetAdministrationAAS-descriptors - or AAS descriptor - in AAS vocabulary) with asset IDs
```

Reviewer: There is not "assetId" attribute, there is the globalAssetId and there are specificAssetIds.

Author: part 2 of the AAS spec, p. 113, bullet point 13 introduces the term as the union of specficAssetIds and globalAssetIds.

>   including at least one that's known at both, the supplier and customer.

Reviewer: ```suggestion
  including at least one asset ID that's known at both sides, the supplier and customer side.
```

- Submodels

>     - registered in a DSP-conformant connector to the Dataspace.

Reviewer: see before: DSP-conformant: whatever this means and where is it described?

    - registered to the correct AAS-descriptor in the DTR.

>     - Serving arbitrary payloads from the `/$value` endpoint, in accordance with their semantic definition.

Reviewer: ```suggestion
    - Serving payloads from the `/$value` endpoint, in accordance with their semantic definition.
```
